state-sync/inter-component/mempool-notifier/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+        if self.notification_receiver_taken() {
+            return Err(Error::NotificationReceiverAlreadyTaken);
+        }
+
+        let (notification_sender, notification_receiver) =
+            mpsc::channel(MEMPOOL_NOTIFIER_CHANNEL_SIZE);
+
+        self.notification_sender = Some(notification_sender);
+        Ok(notification_receiver)
+    }
+
+    fn notification_receiver_taken(&self) -> bool {
+        self.notification_sender.is_some()
+    }
+}
+
+#[async_trait]
+impl MempoolNotifierInterface for MempoolNotifier {
+    async fn notify_commit(
+        &mut self,
+        transactions: Vec<Transaction>,
+        block_timestamp_usecs: u64,
+        notification_timeout_ms: u64,
+    ) -> Result<(), Error> {
+        // Verify the notification receiver has already been taken
+        let notification_sender = if let Some(notification_sender) = &mut self.notification_sender {
+            notification_sender
+        } else {
+            return Err(Error::NotificationReceiverNotTaken);
+        };
+
+        // Get only user transactions from committed transactions
+        let user_transactions: Vec<&Transaction> = transactions
+            .iter()
+            .filter(|transaction| matches!(transaction, Transaction::UserTransaction(_)))
+            .collect();
+
+        // Only send a notification if user transactions have been committed
+        if user_transactions.is_empty() {
+            return Ok(());
+        }
+
+        // Construct a oneshot channel to receive a mempool response
+        let (callback, callback_receiver) = oneshot::channel();
+        let commit_notification = MempoolCommitNotification {
+            transactions,
+            block_timestamp_usecs,
+            callback,
+        };
+
+        // Send the notification to mempool
+        if let Err(error) = notification_sender.try_send(commit_notification) {
+            return Err(Error::CommitNotificationError(format!(
+                "Failed to notify mempool of committed transactions! Error: {:?}",
+                error
+            )));
+        }
+
+        // Handle any responses or a timeout
+        if let Ok(response) = timeout(
+            Duration::from_millis(notification_timeout_ms),
+            callback_receiver,
+        )
+        .await
+        {
+            match response {
+                Ok(Ok(MempoolCommitResponse::Success)) => Ok(()),
+                Ok(Err(error)) => Err(Error::UnexpectedErrorEncountered(format!(
+                    "Unexpected response from mempool! Error: {:?}",
+                    error
+                ))),
+                Err(error) => Err(Error::UnexpectedErrorEncountered(format!("{:?}", error))),
+            }
+        } else {
+            Err(Error::TimeoutWaitingForMempool)
+        }
+    }
+}
+
+/// A notification for newly committed transactions sent by state sync to mempool.
+#[derive(Debug)]
+pub struct MempoolCommitNotification {
+    pub transactions: Vec<Transaction>,
```

Comment:
curious why it's Transaction instead of CommittedTransaction like CommitNotification? it's not part of the interface, more like mempool specific details

Reply:
I'm not sure myself, but I'm assuming it's comparing the `Transaction` from the `MempoolCommitNotification` to the `Transaction` in mempool.